worldedit-core/src/main/java/com/sk89q/worldedit/world/World.java

```diff
@@ -107,17 +108,54 @@
      * @param notifyAndLight true to to notify and light
      * @return true if the block was successfully set (return value may not be accurate)
      */
-    <B extends BlockStateHolder<B>> boolean setBlock(BlockVector3 position, B block, boolean notifyAndLight) throws WorldEditException;
+    @Deprecated
+    default <B extends BlockStateHolder<B>> boolean setBlock(BlockVector3 position, B block, boolean notifyAndLight) throws WorldEditException {
+        return setBlock(position, block, notifyAndLight ? SideEffectApplier.defaults() : SideEffectApplier.none());
+    }
+
+    /**
+     * Similar to {@link Extent#setBlock(BlockVector3, BlockStateHolder)} but a
+     * {@code notifyAndLight} parameter indicates whether adjacent blocks
+     * should be notified that changes have been made and lighting operations
+     * should be executed.
+     *
+     * <p>If it's not possible to skip lighting, or if it's not possible to
+     * avoid notifying adjacent blocks, then attempt to meet the
+     * specification as best as possible.</p>
+     *
+     * <p>On implementations where the world is not simulated, the
+     * {@code notifyAndLight} parameter has no effect either way.</p>
+     *
+     * @param position position of the block
+     * @param block block to set
+     * @param sideEffectApplier which side effects to perform
+     * @return true if the block was successfully set (return value may not be accurate)
+     */
+    <B extends BlockStateHolder<B>> boolean setBlock(BlockVector3 position, B block, SideEffectApplier sideEffectApplier) throws WorldEditException;
+
+    /**
+     * Notifies the simulation that the block at the given location has
+     * been changed and it must be re-lighted (and issue other events).
+     *
+     * @param position position of the block
+     * @param previousType the type of the previous block that was there
+     * @return true if the block was successfully notified
+     */
+    @Deprecated
+    default boolean notifyAndLightBlock(BlockVector3 position, BlockState previousType) throws WorldEditException {
+        return notifyBlock(position, previousType, SideEffectApplier.defaults());
+    }
 
     /**
      * Notifies the simulation that the block at the given location has
      * been changed and it must be re-lighted (and issue other events).
      *
      * @param position position of the block
      * @param previousType the type of the previous block that was there
+     * @param sideEffectApplier which side effects to perform
      * @return true if the block was successfully notified
      */
-    boolean notifyAndLightBlock(BlockVector3 position, BlockState previousType) throws WorldEditException;
+    boolean notifyBlock(BlockVector3 position, BlockState previousType, SideEffectApplier sideEffectApplier) throws WorldEditException;
```

Comment:
I don't like the name `notifyBlock` -- if it is really notifying the simulation, we should name it `sendBlockChange` or something. Personally I prefer the name I had, `applySideEffects`.

Reply:
To be clear, my issue with the current name is that it sounds like it does something with the block itself, which isn't really true, it's about the change at the location.